src/Sign.SignatureProviders.TrustedSigning/TrustedSigningService.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE.txt file in the project root for more information.
+
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using Azure;
+using Azure.CodeSigning;
+using Azure.CodeSigning.Models;
+using Azure.Core;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Sign.Core;
+
+namespace Sign.SignatureProviders.TrustedSigning
+{
+    internal sealed class TrustedSigningService : ISignatureAlgorithmProvider, ICertificateProvider, IDisposable
+    {
+        private static readonly SignRequest _emptyRequest = new(SignatureAlgorithm.RS256, new byte[32]);
+
+        private readonly CertificateProfileClient _client;
+        private readonly string _accountName;
+        private readonly string _certificateProfileName;
+        private readonly ILogger<TrustedSigningService> _logger;
+        private readonly SemaphoreSlim _mutex = new(1);
+        private X509Certificate2? _publicKey;
+
+        public TrustedSigningService(
+            IServiceProvider serviceProvider,
+            TokenCredential tokenCredential,
+            Uri endpointUrl,
+            string accountName,
+            string certificateProfileName)
+        {
+            ArgumentNullException.ThrowIfNull(serviceProvider, nameof(serviceProvider));
+            ArgumentNullException.ThrowIfNull(tokenCredential, nameof(tokenCredential));
+            ArgumentNullException.ThrowIfNull(endpointUrl, nameof(endpointUrl));
+            ArgumentException.ThrowIfNullOrEmpty(accountName, nameof(accountName));
+            ArgumentException.ThrowIfNullOrEmpty(certificateProfileName, nameof(certificateProfileName));
+
+            _client = new CertificateProfileClient(tokenCredential, endpointUrl);
+            _accountName = accountName;
+            _certificateProfileName = certificateProfileName;
+
+            _logger = serviceProvider.GetRequiredService<ILogger<TrustedSigningService>>();
+        }
+
+        public void Dispose()
+        {
+            _mutex.Dispose();
+            _publicKey?.Dispose();
+            GC.SuppressFinalize(this);
+        }
+
+        public async Task<X509Certificate2> GetCertificateAsync(CancellationToken cancellationToken)
+        {
+            if (_publicKey is not null)
+            {
+                return new X509Certificate2(_publicKey);
```

Comment:
Can this be cached/reused so that multiple requests don't new up additional certificate objects

Reply:
When I do that I get the following error:

```
fail: Sign.Core.ISigner[0]
      The chain context handle is invalid. (Parameter 'certificate')
      System.ArgumentException: The chain context handle is invalid. (Parameter 'certificate')
         at System.Security.Cryptography.X509Certificates.X509Chain.Build(X509Certificate2 certificate, Boolean throwOnException)
         at AzureSign.Core.AuthenticodeKeyVaultSigner..ctor(AsymmetricAlgorithm signingAlgorithm, X509Certificate2 signingCertificate, HashAlgorithmName fileDigestAlgorithm, TimeStampConfiguration timeStampConfiguration, X509Certificate2Collection additionalCertificates)
         at Sign.Core.AzureSignToolSigner.SignAsync(IEnumerable`1 files, SignOptions options) in D:\PullRequests\dotnet\sign\src\Sign.Core\DataFormatSigners\AzureSignToolSigner.cs:line 96
         at Sign.Core.AggregatingSigner.SignAsync(IEnumerable`1 files, SignOptions options) in D:\PullRequests\dotnet\sign\src\Sign.Core\DataFormatSigners\AggregatingSigner.cs:line 204
         at Sign.Core.AggregatingSigner.SignAsync(IEnumerable`1 files, SignOptions options) in D:\PullRequests\dotnet\sign\src\Sign.Core\DataFormatSigners\AggregatingSigner.cs:line 92
         at Sign.Core.Signer.<>c__DisplayClass3_0.<<SignAsync>b__0>d.MoveNext() in D:\PullRequests\dotnet\sign\src\Sign.Core\Signer.cs:line 155
      --- End of stack trace from previous location ---
         at System.Threading.Tasks.Parallel.<>c__53`1.<<ForEachAsync>b__53_0>d.MoveNext()
      --- End of stack trace from previous location ---
         at Sign.Core.Signer.SignAsync(IReadOnlyList`1 inputFiles, String outputFile, FileInfo fileList, DirectoryInfo baseDirectory, String applicationName, String publisherName, String description, Uri descriptionUrl, Uri timestampUrl, Int32 maxConcurrency, HashAlgorithmName fileHashAlgorithm, HashAlgorithmName timestampHashAlgorithm) in D:\PullRequests\dotnet\sign\src\Sign.Core\Signer.cs:line 84
```

That is happening because we are doing `using (X509Certificate2 certificate = await _certificateProvider.GetCertificateAsync())` inside `AzureSignToolSigner.SignAsync`.

Reply:
seems like we shouldn't be disposing that cert then, at least not at that scope?

Reply:
This feedback and others like it is out of scope for this PR.  @clairernovotny, your concerns are based on an existing design, which @dlemstra is following correctly.

It is a best practice to dispose of `IDisposable` instances for which you've taken ownership of and you no longer need.  @dlemstra's changes follow existing patterns.  The scope of this PR should not include the suggested design change.

In the future, we could redesign signers to reuse `X509Certificate2` instances as much as possible.  For example, an `X509Certificate2` instance could be passed to `IDataFormatSigner.SignAsync(...)`.  The instance would be owned and disposed of at a more outer scope.  There may be alternatives.

Also, while performance is important, so are many other things, like maintainability, readability, etc.  Measuring the performance benefit of such a change would be very valuable in cost/benefit decisions.  Reusing `X509Certificate2` instances probably improves overall performance (by virtue of fewer allocations and GC's), but without measurements, it's hard to say if it'll be a significant benefit.

For example, for signing a NuGet package and the assemblies inside of it, it looks like --- from a cursory glance --- we'll instantiate and dispose of 2 `X509Certificate2` objects in the service of `GetCertificateAsync(...)` calls.  A design change that would reuse `X509Certificate2` instances would bring that down from 2 to 1.  On the surface (read:  without actual measurements), all of that redesign work to eliminate 1 instantiation seems like not a good use of perf optimization time.  IMO, a better use of time would be to profile Sign CLI end-to-end and look for low-hanging fruit.

@clairernovotny, please open a new issue if you feel strongly about this.